libpermian/plugins/anaconda_webui/__init__.py

```diff
@@ -0,0 +1,498 @@
+        self.vm_semaphore.acquire()
+        if self.canceled:
+            return
+        self.reportResult(Result('started', None, False))
+        self._start_vm()
+
+        # Wait for WebUI to start
+        self._wait_for_ip()
+        self._wait_for_webui()
+
+    def execute(self):
+        if self.canceled:
+            return
+        self.log(f'Running test {self.test_case_name}', show=True)
+        self.reportResult(Result('running', None, False))
+
+        cmd = [self.test_file, self.test_case_name, 
+               '--browser', f'{self.vm_ip}:{self.port_webui}',
+               '--machine', f'{self.vm_ip}:{self.port_ssh}']
+        
+        self.log('Running: ' + ' '.join(cmd))
+        if self.test_repo_name:
+            # Test in separate repo
+            test_env = {'WEBUI_TEST_DIR': os.path.abspath(os.path.join(self.webui_dir, 'test'))} # tells the test where is anaconda repo
+            if self.use_container:
+                cont_cwd = self.test_workdir.replace(self.temp_dir, '/root/workdir')
+                test_env['WEBUI_TEST_DIR'] = test_env['WEBUI_TEST_DIR'].replace(self.temp_dir, '/root/workdir')
+                process = self.container.exec(cmd, self.temp_dir, env=test_env, cwd=cont_cwd, log_error=False)
+            else:
+                process = subprocess.run(cmd, cwd=self.test_workdir, env=test_env, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+
+        else:
+            # Test in anaconda repo
+            if self.use_container:
+                cont_cwd = self.test_workdir.replace(self.temp_dir, '/root/workdir')
+                process = self.container.exec(cmd, self.temp_dir, cwd=cont_cwd, log_error=False)
+            else:
+                process = subprocess.run(cmd, cwd=self.test_workdir, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+
+        self.log(process.stdout.decode(), 'output')
+        self.log('Test finished', show=True)
+
+        if process.returncode != 0:
+            self.reportResult(Result('complete', 'FAIL', True))
+        else:
+            self.reportResult(Result('complete', 'PASS', True))            
+
+    def teardown(self):
+        self.log('Teardown', show=True)
+        self._collect_test_logs()
+        # Kill VM        
+        self._virsh_call(['destroy', self.vm_name])
+        # Remove VM
+        self._virsh_call(['undefine', self.vm_name, '--remove-all-storage'], check=True)
+        
+        self.vm_semaphore.release()
+
+        with self.setup_lock:
+            # If this instance is the last one using temp dir -> remove it
+            self.instances[self.git_anaconda_branch].remove(self)
+            if len(self.instances[self.git_anaconda_branch]) == 0:
+                self.temp_dirs[self.git_anaconda_branch].cleanup()
+                
+                del self.instances[self.git_anaconda_branch]
+                # Remove container image, if this was the last anacoda-webui workflow
+                if self.instances == dict():
```

Comment:
Instead of checking the `self.instances` equals empty dict, the more python way is to check if "it's empty" by expression `not self.instances` which is a expression that can be used for majority of containers and iterables (as well as None).